service/history/mutableStateBuilder.go

```diff
@@ -1973,120 +1619,23 @@ func (e *mutableStateBuilder) AddDecisionTaskCompletedEvent(
-		startedEvent := e.hBuilder.AddTransientDecisionTaskStartedEvent(scheduledEvent.GetEventId(), decision.RequestID,
-			request.GetIdentity(), decision.StartedTimestamp)
-		startedEventID = startedEvent.GetEventId()
-	}
-	// Now write the completed event
-	event := e.hBuilder.AddDecisionTaskCompletedEvent(scheduleEventID, startedEventID, request)
-
-	e.afterAddDecisionTaskCompletedEvent(event, maxResetPoints)
-	return event, nil
+	return e.dtm.AddDecisionTaskCompletedEvent(scheduleEventID, startedEventID, request, maxResetPoints)
 }
 
-func (e *mutableStateBuilder) ReplicateDecisionTaskCompletedEvent(
-	event *workflow.HistoryEvent,
-) error {
-
-	e.beforeAddDecisionTaskCompletedEvent()
-	e.afterAddDecisionTaskCompletedEvent(event, math.MaxInt32)
-	return nil
+func (e *mutableStateBuilder) ReplicateDecisionTaskCompletedEvent(event *workflow.HistoryEvent) error {
+	return e.dtm.ReplicateDecisionTaskCompletedEvent(event)
 }
 
-func (e *mutableStateBuilder) AddDecisionTaskTimedOutEvent(
-	scheduleEventID int64,
-	startedEventID int64,
-) (*workflow.HistoryEvent, error) {
-
-	opTag := tag.WorkflowActionDecisionTaskTimedOut
-	if err := e.checkMutability(opTag); err != nil {
-		return nil, err
-	}
-
-	dt, ok := e.GetDecisionInfo(scheduleEventID)
-	if !ok || dt.StartedID != startedEventID {
-		e.logger.Warn(mutableStateInvalidHistoryActionMsg, opTag,
-			tag.WorkflowEventID(e.GetNextEventID()),
-			tag.ErrorTypeInvalidHistoryAction,
-			tag.WorkflowScheduleID(scheduleEventID),
-			tag.WorkflowStartedID(startedEventID))
-		return nil, e.createInternalServerError(opTag)
-	}
-
-	var event *workflow.HistoryEvent
-	// Avoid creating new history events when decisions are continuously timing out
-	if dt.Attempt == 0 {
-		event = e.hBuilder.AddDecisionTaskTimedOutEvent(scheduleEventID, startedEventID, workflow.TimeoutTypeStartToClose)
-	}
-
-	if err := e.ReplicateDecisionTaskTimedOutEvent(workflow.TimeoutTypeStartToClose); err != nil {
-		return nil, err
-	}
-	return event, nil
+func (e *mutableStateBuilder) AddDecisionTaskTimedOutEvent(scheduleEventID int64, startedEventID int64) (*workflow.HistoryEvent, error) {
+	return e.dtm.AddDecisionTaskTimedOutEvent(scheduleEventID, startedEventID)
 }
 
-func (e *mutableStateBuilder) ReplicateDecisionTaskTimedOutEvent(
-	timeoutType workflow.TimeoutType,
-) error {
-
-	incrementAttempt := true
-	// Do not increment decision attempt in the case of sticky timeout to prevent creating next decision as transient
-	if timeoutType == workflow.TimeoutTypeScheduleToStart {
-		incrementAttempt = false
-	}
-	e.FailDecision(incrementAttempt)
-	return nil
+func (e *mutableStateBuilder) ReplicateDecisionTaskTimedOutEvent(timeoutType workflow.TimeoutType) error {
+	return e.dtm.ReplicateDecisionTaskTimedOutEvent(timeoutType)
 }
 
-func (e *mutableStateBuilder) AddDecisionTaskScheduleToStartTimeoutEvent(
-	scheduleEventID int64,
-) (*workflow.HistoryEvent, error) {
-
-	opTag := tag.WorkflowActionDecisionTaskTimedOut
```

Comment:
i believe the following checks should be kept in mutable state
```go
opTag := tag.WorkflowActionDecisionTaskTimedOut
if err := e.checkMutability(opTag); err != nil {
	return nil, err
}
```

Reply:
same applies to all places

Reply:
I like the property that all the logic of knowing how to update a decision task in accordance with the current state of the workflow lives in one file. Why do you prefer split this logic into two places?

Reply:
the mutability check should exist in mutable state, which in general controls all logic
decision state machine should be the logic maintaining the decision part, nothing else

Reply:
I would agree if the file I was adding was a pure state machine. But as we spoke about offline on Thursday it is not this. Really all this file is a grouping of related methods from mutableStateBuilder. 

If we only want the logic of decision in this new file then we also need to remove things like: checking buffered events and writing to hBuilder. 

In short I think decisionInfo updates are too tightly duplicated with other parts of mutableStateBuilder to pull it out cleanly. So instead I think just all the methods related to decisions should be pulled out.